concurrency/MageUI.java

```diff
@@ -0,0 +1,103 @@
+package concurrency;
+
+import javax.swing.*;
+import java.awt.*;
+import java.util.EnumMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class MageUI {
+
+    private final Map<MageComponents, Component> ui = new EnumMap<>(MageComponents.class);
+    private final Map<MageComponents, Object> sync = new EnumMap<>(MageComponents.class);
+
+    public JButton getButton(MageComponents name) throws InterruptedException {
+        //System.out.println("request for " + name);
+        Object buttonSync;
+        synchronized (ui) {
+            if (ui.containsKey(name)) {
+                //System.out.println("clicking " + name);
+                return (JButton) ui.get(name);
+            } else {
+                buttonSync = new Object();
+                sync.put(name, buttonSync);
+            }
+        }
+
+        synchronized (buttonSync) {
+            //System.out.println("waiting " + name + " to be created");
+            buttonSync.wait();
+            //System.out.println(name + "was created");
+            if (!ui.containsKey(name)) {
+                throw new IllegalStateException("Component wasn't initialized. This should not happen.");
+            }
+            return (JButton) ui.get(name);
+        }
+
+    }
+
+    public Component getComponent(MageComponents name) throws InterruptedException {
+        Object componentSync;
+        synchronized (ui) {
+            if (ui.containsKey(name)) {
+                return ui.get(name);
+            } else {
+                componentSync = new Object();
+                sync.put(name, componentSync);
+            }
+        }
+
+        synchronized (componentSync) {
+            componentSync.wait();
+            if (!ui.containsKey(name)) {
+                throw new IllegalStateException("Component wasn't initialized. This should not happen.");
+            }
+            return ui.get(name);
+        }
+
+    }
+
+    public void addButton(MageComponents name, JButton button) {
+        synchronized (ui) {
+            //System.out.println("added " + name);
+            ui.put(name, button);
+            if (sync.containsKey(name)) {
+                synchronized (sync.get(name)) {
+                    //System.out.println("notifyAll - " + name);
+                    sync.get(name).notifyAll();
+                }
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 61:70, 80:87.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.